### PR TITLE
Added support for approve pattern

### PR DESCRIPTION
**Pull request by amiromayer**
@Dobrokhvalov Implemented issue #87

### CODE REVIEW

File: contracts/contracts/Factory.sol [hunk 1]
```diff
@@ -32,6 +35,37 @@ contract Factory is Storage, CloneFactory {
...
     }
 
+    /**
+    * @dev Function to get total amount of tokens available for proxy contract
+    * @param _proxy Address of proxy contract
+    * @param _sender Address of lindkrop sender
+    * @param _token Token address, address(0) for ETH
+    * @return Total amount available
+    */
+    function getBalance(address _proxy, address _sender, address _token) public view returns (uint) {
```
Comment (Dobrokhvalov) [inline contracts/contracts/Factory.sol]: can't we use `deployed[_sender]` to get _proxy address?

File: contracts/contracts/Factory.sol [hunk 2]
```diff
@@ -32,6 +35,37 @@ contract Factory is Storage, CloneFactory {
...
+
+        if (_token == address(0)) {
+            return _proxy.balance;
+        }
+        else {
+            uint allowance = IERC20(_token).allowance(_sender, _proxy);
+            uint balance = IERC20(_token).balanceOf(_proxy);
+            return allowance + balance;
```
Comment (Dobrokhvalov) [inline contracts/contracts/Factory.sol]: use SafeMath to protect from overflows - https://ethereumdev.io/safemath-protect-overflows/

File: contracts/contracts/Linkdrop.sol [hunk 3]
```diff
@@ -158,6 +158,24 @@ contract Linkdrop is ILinkdrop, Common {
...
+    */
+    function getBalance(address _token) public view returns (uint) {
+
+        if (_token == address(0)) {
+            return address(this).balance;
+        }
+        else {
+            uint allowance = IERC20(_token).allowance(SENDER, address(this));
+            uint balance = IERC20(_token).balanceOf(address(this));
+            return allowance + balance;
```
Comment (Dobrokhvalov) [inline contracts/contracts/Linkdrop.sol]: use SafeMath

File: contracts/contracts/Linkdrop.sol [hunk 4]
```diff
@@ -169,16 +187,45 @@ contract Linkdrop is ILinkdrop, Common {
...
+                if (balance >= _amount) 
+                    fromProxyAmount = _amount;
+                else 
+                    fromProxyAmount = balance;
+
+                // Transfer tokens from proxy balance
+                IERC20(_token).transfer(_receiver, fromProxyAmount); 
+                
+                // Transfer rest funds from sender's balance
+                uint rest = _amount - fromProxyAmount;
```
Comment (Dobrokhvalov) [inline contracts/contracts/Linkdrop.sol]: use SafeMath

File: contracts/contracts/Factory.sol (hunk 1 above)
Comment (amiromayer) [reply contracts/contracts/Factory.sol]: No, indeed we need this function when proxy is not deployed yet, so we can't know proxy address onchain

Comment (amiromayer) [reply contracts/contracts/Factory.sol]: When checking claim params from factory for instance

Comment (Dobrokhvalov) [reply contracts/contracts/Factory.sol]: ok

**Pull request merged**